### PR TITLE
Module4 homework

### PR DESCRIPTION
**Pull request by pwrice**

### CODE REVIEW

File: module4/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+properties, methods
+
+i) A class, struct, or enum that adopts a protocol is said to _______________ to the protocol
+
+conform
+
+j) A lazy property is calculated only when ________________________.
+
+it is accessed
```
Comment (waltertyree) [inline module4/README.md:42]: When it is _first_ accessed, not every time. A calculated property is every time.

File: module4/Homework4Playground.playground/Contents.swift [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+  print("calling average() overload nonOptional parameter")
+  average(array)
+}
+
+// average with default parameter
+func average(array: [Int]? = firstArry) {
+  print("calling average() overload with default parameter")
+  average(array)
+}
+
```
Comment (waltertyree) [inline module4/Homework4Playground.playground/Contents.swift:34]: these just call the original. What is also common is to replace the original to extend the function. For example
````
func average(_ array: [Double]) {
let total = array.reduce(0.0, +)
print("The average of the values in the array is \(total/Double(array.count))")
}
````

File: module4/Homework4Playground.playground/Contents.swift [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+// Is this what is meant?
+var numsBy4: [Int] = Array(1...25).map { $0 * 4 - 2 }
+print(numsBy4)
+
+// Or perhaps this?
+numsBy4 = Array(1...25).map { $0 * 4 }
+print(numsBy4)
+
+// Or perhaps this?
+numsBy4 = Array(stride(from:2, through: 100, by:4))
```
Comment (waltertyree) [inline module4/Homework4Playground.playground/Contents.swift:103]: Kind of this. Using stride.

File: module4/Homework4Playground.playground/Contents.swift [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+print(numsBy2)
+
+//Create an array of Int called numsBy4 with values starting at 2 through 100, by 4.
+
+// Is this what is meant?
+var numsBy4: [Int] = Array(1...25).map { $0 * 4 - 2 }
+print(numsBy4)
+
+// Or perhaps this?
+numsBy4 = Array(1...25).map { $0 * 4 }
```
Comment (waltertyree) [inline module4/Homework4Playground.playground/Contents.swift:99]: No, not this. This doesn't produce the correct array

File: module4/Homework4Playground.playground/Contents.swift [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+  array.filter({ $0 % 2 == 0})
+}
+
+let evenNums = evenNumbersArray(nums)
+print("evenNums \(evenNums)")
+
+//- Create a function called sumOfArray that takes a parameter of [Int?] and returns an Int. The function should return the sum of the array values passed that are not nil. Call the function passing the numsWithNil array, and print out the results.
+
+func sumOfArray(_ array: [Int?]) -> Int {
+  array.compactMap({ $0 }).reduce(0, { $0 + $1 })
```
Comment (waltertyree) [inline module4/Homework4Playground.playground/Contents.swift:121]: You get the right answer, but I'd normally expect `reduce` to look like this:
````
  array.compactMap({ $0 }).reduce(0, +)
````

File: module4/Homework4Playground.playground/Contents.swift [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+let circle = Circle(radius: 3)
+print("circle area: \(circle.calculateArea())")
+
+let rect = Rectangle(width: 3, height: 4)
+print("rect area: \(rect.calculateArea())")
+
+//Next, extend the protocol Shape to add a new method called calculateVolume() -> Double.
+
+
+// Interesting - it seems you can extend a protocol w/out defining a default implementation. I guess that makes sense?
```
Comment (waltertyree) [inline module4/Homework4Playground.playground/Contents.swift:187]: ? You do provide a default implementation for your new func. I don't understand your comment.

**Review by waltertyree — APPROVED** [review]
I added some comments to add context.

File: module4/Homework4Playground.playground/Contents.swift (hunk 6 above)
Comment (pwrice) [reply module4/Homework4Playground.playground/Contents.swift:187]: typo: 
I meant:
"it seems you * can't * extend a protocol w/out defining a default implementation."

Which was surprising to me when I wrote this. But it makes sense in that, if others have already implemented the protocol, then you cant add a function that they then must go implement to conform.

**Pull request merged**